### PR TITLE
Add org.multimc.MultiMC (MultiMC)

### PR DESCRIPTION
**Pull request by joshua-stone**
This initial pull request creates an unofficial distribution of MultiMC that includes the latest LTS release of OpenJDK to ensure it can launch successfully on first run.

### CODE REVIEW

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

File: org.multimc.MultiMC.appdata.xml
Comment (hfiguiere) [inline org.multimc.MultiMC.appdata.xml]: the proper SPDX ID is "Apache-2.0"

source https://spdx.org/licenses/

(hence the failure)

Comment (joshua-stone) [reply org.multimc.MultiMC.appdata.xml]: Thank you, I'll try rebuilding with this fix!

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

File: org.multimc.MultiMC.yml
Comment (LeandroStanger) [inline org.multimc.MultiMC.yml:46]: New appdata files should use metainfo instead. https://www.freedesktop.org/software/appstream/docs/chap-Quickstart.html

File: org.multimc.MultiMC.yml
Comment (barthalion) [inline org.multimc.MultiMC.yml:20]: I'd consider bundling regular org.freedesktop.Sdk.Extension.openjdk just for the sake of it, as I assume it's something people would like to have.

File: org.multimc.MultiMC.yml
Comment (barthalion) [inline org.multimc.MultiMC.yml:38]: That should be kept around, it's split automatically to `Debug` extension.

File: org.multimc.MultiMC.yml
Comment (barthalion) [inline org.multimc.MultiMC.yml:10]: I guess it doesn't actually work natively on Wayland?

File: org.multimc.MultiMC.appdata.xml
Comment (barthalion) [inline org.multimc.MultiMC.appdata.xml:35]:
```suggestion
    <release version="0.6.11" date="2020-03-28" />
```

File: org.multimc.MultiMC.appdata.xml
Comment (barthalion) [inline org.multimc.MultiMC.appdata.xml:1]: Apart from upstream not being a fan of flatpak/snap/etc, please try to submit it upstream as it's also useful to regular Linux distributions.

File: org.multimc.MultiMC.yml
Comment (joshua-stone) [reply org.multimc.MultiMC.yml:10]: Which part of the program isn't working natively in Wayland?

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

File: org.multimc.MultiMC.metainfo.xml
Comment (bilelmoussaoui) [inline org.multimc.MultiMC.metainfo.xml:32]: A screenshot can only have one image tag.

Comment (bilelmoussaoui) [reply org.multimc.MultiMC.metainfo.xml:32]: So you need multiple screenshot tag here

File: org.multimc.MultiMC.yml
Comment (bilelmoussaoui) [inline org.multimc.MultiMC.yml:2]:
```suggestion
```

File: org.multimc.MultiMC.yml
Comment (bilelmoussaoui) [reply org.multimc.MultiMC.yml:22]: This wasn't resolved

File: org.multimc.MultiMC.metainfo.xml
Comment (joshua-stone) [reply org.multimc.MultiMC.metainfo.xml]: I think the image tags should look better now.

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

File: org.multimc.MultiMC.yml
Comment (barthalion) [reply org.multimc.MultiMC.yml:21]: It was? There are 3 JDK versions now.

File: org.multimc.MultiMC.yml
Comment (barthalion) [inline org.multimc.MultiMC.yml:67]: FYI there's also desktop-file-edit that allows things like this

**Comment by barthalion** [comment]
Hey @peterix, I know you're not a fan of flatpak/snap, but would you like to have write access to the Flathub repository once it's created?

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by peterix** [comment]
I want this thing deleted.

**Comment by peterix** [comment]
At no point should it be allowed to exist.

**Review by peterix — CHANGES_REQUESTED** [review]
You are violating the license of MultiMC.

**Comment by barthalion** [comment]
You published the code under Apache license. There's no license violation here.

**Comment by peterix** [comment]
You made changes to it and did not remove the branding.

This is *NOT* MultiMC.

**Comment by peterix** [comment]
I really do not have the patience to argue with you about this. Already did with @joshua-stone, for hours.

No, I do not want this to exist. Yes, I said why, many, MANY times, in way too much detail.

@joshua-stone doesn't care and did *this* anyway.

No means no.

**Comment by barthalion** [comment]
The code is changed to adapt some paths to JDKs and there's no functional changes. I don't get why you're being so unreasonable; someone is actually investing their free time here to make your application available somewhere and ready to handle all the related maintenance.

The application description makes it that in no way this package is affiliated with MultiMC project – even though it's not technically required by the license, even if MultiMC was a registered trademark.

**Comment by peterix** [comment]
OK. So you have no respect for anyone either.

**Comment by barthalion** [comment]
I have mentioned you purely out of respect but sadly, I don't think we're going to get anywhere this way

**Comment by jikuja** [comment]
> his distribution is not verified by, affiliated with, or supported by the MultiMC team.

Where are your bug tracker and other support channels for problems with the bundled app?

**Comment by peterix** [comment]
I don't want this to go anywhere. It should not exist. I will not support it. It should not use the MultiMC name. And I do not want the maintenance or support of this to ever be forced upon me or any of the people who ever contributed to the project.

**Comment by peterix** [comment]
Now kidly remove it.

**Comment by peterix** [comment]
And it is you being unreasonable, not willing to take 'no' as an answer.

**Comment by barthalion** [comment]
At what length would we need to go to make it work? Just removing any URLs that would point to official communication channels, making it even more clear it's not official (even so by adding a welcome window stating this and pointing what is officially supported), or is a full rename/fork the only satisfying conclusion here?

**Comment by Janrupf** [comment]
From the MultiMC repo itself:
> Part of the reason for using the Apache license is we don't want people using the "MultiMC" name when redistributing the project. This means people must take the time to go through the source code and remove all references to "MultiMC", including but not limited to the project icon and the title of windows, (no MultiMC-fork in the title).
> 
> Apache covers reasonable use for the name - a mention of the project's origins in the About dialog and the license is acceptable. However, it should be abundantly clear that the project is a fork without implying that you have our blessing.

I think that should make it pretty clear that a full rename/fork is required redistribute the project in compliance with the license.

**Comment by barthalion** [comment]
I've read the README. I'm asking it again here, but thanks.

**Comment by peterix** [comment]
There are plenty of reasons for this, btw.

### MultiMC handles user credentials and enforces DRM on behalf of Mojang/Microsoft.
I had to ask the snap people to take down one of those because people put up a cracked version on the snap repos. Thankfully, they had the decency to take it down and I didn't have to figure out how to follow up on that.

I'm not saying that you are doing something wrong in this regard, but I do not have the time or inclination to audit what you do on an ongoing basis. I'm in the same relationship with Mojang/MS - very unofficial.

And yeah, there are builds with auth/drm disabled or cracked copies of MultiMC floatong around the internet. I wish there weren't, because it's a liability.

### The game is pretty complicated to get running correctly
The mods and variability in user systems is a bad enough problem without unofficial builds floating around.  It's OK if you mess with the code as an end user and you generally have some idea about what you are doing. ... but if you push that to others and call it MultiMC while leaving me to deal with the support issues, that's sersiously NOT OK.

Currently, MultiMC and the Mojang launcher do not bundle any Java. MultiMC should follow what Mojang is doing in this regard. Different versions of Java will be bundled in the future based on the needs of the game, not based on what you think was appropriate to throw in.

### The license is used to enforce the above

I've considered removing all branding, infrastructure URLs, and links to support places from the codebase before. Might have to do it sooner or later, for multiple reasons. You are not the first to want to redistribute it, and with the name and logo being there by default and not easy to change in one place, it's not as easy to maintain such a rebranded fork.

While the name and logo are not registered as a trademark, I believe they are still protected under common law (IANAL, and I don't particularly feel like paying lawyers unless I absolutely have to).

### Future issues with MS authentication

One major upcoming concern is that Microsoft is getting involved with third party launchers. You will likely not be able to build a full-featured version of the launcher without registering the application with Microsoft in the future.

So, you **may** not be able to make working builds in the future without becoming responsible for user safety and privacy (having a formal relationship with Mojang/MS, implementing MSA correctly, etc), or without actively trying to impersonate me.
I am not OK with being liable for your actions.

---

Ultimately, I've been doing this for years and feel responsible. From my POV, you are making flatpak my problem against my will, while walking into a bunch of problems that you are not ready to handle.

So. My answer is a decisive NO.

If you really want to deal with maintaining a Minecraft launcher, go ahead. Fork it completely, infrastructure included. Please don't half-ass it. I don't want even a hint of implication that I am involved in whatever you are doing.

**Comment by barthalion** [comment]
> Different versions of Java will be bundled in the future based on the needs of the game, not based on what you think was appropriate to throw in.

Yeah, I realize now by looking at the docs that the only built-in Java should have been 8.

> From my POV, you are making flatpak my problem against my will, while walking into a bunch of problems that you are not ready to handle.

I don't have such a strong view on this – the usual procedure with any non-supported packaging is to just close such bug reports and tell users to complain to whoever packaged the software.

The manifest really doesn't do much more than running CMake after git clone as described at: https://github.com/flathub/flathub/pull/1978/files#diff-06d25db730a0affd31353a80acc5e64389d7ef7cad1898da8da7cd4000a23cdbR51-R58 What we've done in the past for non-trivial software like Chrome/Chromium is putting the package in the beta repository which is not listed on our web store, so only people who know it's there used it for testing; we could always pull the plug for it without swarm of users complaining something has disappeared.

If it's still a no, I leave the decision to @joshua-stone if he wants to follow up with rename.

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
The latest build removes OpenJDK 11 & 15 while leaving OpenJDK 8 to align more closely with build requirements:

https://github.com/MultiMC/MultiMC5/blob/develop/BUILD.md#dependencies

More OpenJDK versions could certainly be added in through extensions, but having them all on first run isn't necessary.

**Comment by joshua-stone** [comment]
bot, build org.multimc.MultiMC

**Comment by joshua-stone** [comment]
This latest build adds some build switches that were found in the stable release PKGBUILD:

https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=multimc5#n55

**Review by joshua-stone — COMMENTED** [review]
> > this distribution is not verified by, affiliated with, or supported by the MultiMC team.
> 
> Where are your bug tracker and other support channels for problems with the bundled app?

Every application maintained on flathub is given its own bug tracker. For example, here is the Obsidian flatpak bug tracker:

https://github.com/flathub/md.obsidian.Obsidian/issues

And here is the upstream Obsidian's bug tracker for general Linux bugs:

https://forum.obsidian.md/search?q=linux%20category%3A7%20order%3Alatest

**Comment by joshua-stone** [comment]
Hey @peterix, my apologies for the late response!

> ### MultiMC handles user credentials and enforces DRM on behalf of Mojang/Microsoft.
> 
> I had to ask the snap people to take down one of those because people put up a cracked version on the snap repos. Thankfully, they had the decency to take it down and I didn't have to figure out how to follow up on that.
>

I'm sorry you've had to deal with cracked clients. I've supported Minecraft since it was in Alpha development, and it can be frustrating to see the MC community being pumped with shady mods and cracked launchers undermining the efforts and legitimacy of those who've put their time into Minecraft and passion projects such as MultiMC.

> I'm not saying that you are doing something wrong in this regard, but I do not have the time or inclination to audit what you do on an ongoing basis. I'm in the same relationship with Mojang/MS - very unofficial.
> 
> And yeah, there are builds with auth/drm disabled or cracked copies of MultiMC floatong around the internet. I wish there weren't, because it's a liability.

Auditability is a very important property in software, and I'm glad that you prioritize it. There are a few ways you can audit flatpak builds:

* [Flathub git repositories](https://github.com/flathub)
* [Flathub's CI system](https://flathub.org/builds/#/)

It would be virtually impossible for someone to sneak in shady patches because it would need approval for merging from a maintainer, and there's an extensive audit trail left behind since it's all publicly visible in git history. For example, this is the commit hash of the latest change I made as of this writing:

https://github.com/flathub/flathub/pull/1978/commits/c253c552523fd8e8e286fe73d073e9e6d0765f1b

Which is found in the Build Properties of the respective build in Flathub's CI:

https://flathub.org/builds/#/builders/32/builds/34376 

The build output can be readily inspected and compared against local builds, which I often do by running:

```
$ flatpak-builder --verbose --install-deps-from=flathub _build --force-clean org.multimc.MultiMC.yml
```

> ### The game is pretty complicated to get running correctly
> 
> The mods and variability in user systems is a bad enough problem without unofficial builds floating around. It's OK if you mess with the code as an end user and you generally have some idea about what you are doing. ... but if you push that to others and call it MultiMC while leaving me to deal with the support issues, that's sersiously NOT OK.

As barthalion pointed out, the expected procedure for supporting a flatpak is that users file bugs on its bug tracker. It's not any different from users having to go to a bug tracker for Debian, Fedora, etc. For example, there's a bug tracker for the [Obsidian flatpak I maintain](https://github.com/flathub/flathub/pull/1978#pullrequestreview-548781955)

In addition to there being a bug tracker, flatpak gives users the tooling they need to extensively debug. This has been useful for when I've had to [help debug Wayland crashes with the Veloren developers](https://github.com/Smithay/wayland-rs/issues/359):

```
[jstone@localhost ~]$ flatpak install net.veloren.veloren net.veloren.veloren.Debug
[jstone@localhost ~]$ flatpak run --command=sh --devel net.veloren.veloren
[📦 net.veloren.veloren ~]$ gdb /app/bin/veloren-voxygen
```

> Currently, MultiMC and the Mojang launcher do not bundle any Java on linux. MultiMC should follow what Mojang is doing in this regard. Different versions of Java will be bundled in the future based on the needs of the game, not based on what you think was appropriate to throw in.
> ### The license is used to enforce the above
 
Out-of-the-box Java support can certainly be moved out of the build for fitting more in line with the goals of Minecraft and MultiMC. The current configuration of [bundling the prepackaged openjdk8](https://github.com/flathub/org.freedesktop.Sdk.Extension.openjdk8) is not intrinsic to flatpak; you can either keep the base application minimal with support libraries installed separately, or bundle whatever you feel is necessary. In this case, it's merely a convenience for first runs because flathub already packages openjdk8, and [whether it's bundled or installed as a shared system library is irrelevant due to ostree's design deduplicating files](https://ostreedev.github.io/ostree/introduction/#atomic-transitions-between-parallel-installable-read-only-filesystem-trees).

> I've considered removing all branding, infrastructure URLs, and links to support places from the codebase before. Might have to do it sooner or later, for multiple reasons. You are not the first to want to redistribute it, and with the name and logo being there by default and not easy to change in one place, it's not as easy to maintain such a rebranded fork.
>
> While the name and logo are not registered as a trademark, I believe they are still protected under common law (IANAL, and I don't particularly feel like paying lawyers unless I absolutely have to).
> ### Future issues with MS authentication

Having to extensively patch MultiMC for a rebrand is certainly not going to be simple, but is necessary for satisfying requirements for a downstream package. If you can, please list which exact parts of the application would need to be patched.

Here's a (probably incomplete) list of what I believe would have to be changed based on a cursory glance:

- App ID in the repository
- Application ID that'll be seen in the desktop environment's launch menu
- Application icon
- Application name in window title bar
- Patreon support URL (I'd personally prefer to leave the URL unchanged, but could change the text to "Support Upstream")
- Application name in debug logs
- Application name in Settings

> One major upcoming concern is that Microsoft is getting involved with third party launchers. You will likely not be able to build a full-featured version of the launcher without registering the application with Microsoft in the future.

> So, you **may** not be able to make working builds in the future without becoming responsible for user safety and privacy (having a formal relationship with Mojang/MS, implementing MSA correctly, etc), or without actively trying to impersonate me.

> I am not OK with being liable for your actions.
 
> Ultimately, I've been doing this for years and feel responsible. From my POV, you are making flatpak my problem against my will, while walking into a bunch of problems that you are not ready to handle.

Whatever issues may impact the flatpak shouldn't be your problem, [as it was already made clear from the very first commit](https://github.com/flathub/flathub/pull/1978/commits/3ec36be53a83e47e607c7cc26397ced5e6e2384d#diff-cf3b11263c408da0a42eb114a55e434c2439900ccf0d4d9e58953a4149dac88bR21) as well as in the [original license](http://www.apache.org/licenses/LICENSE-2.0.html)

One of the key benefits of the open source community is that volunteers can put in improving the quality of software in a transparent, especially if there's a policy in upstreaming patches so that everyone can benefit. That is the relationship I intend to have when maintaining this.

> So. My answer is a decisive NO.
> 
> If you really want to deal with maintaining a Minecraft launcher, go ahead. Fork it completely, infrastructure included. Please don't half-ass it. I don't want even a hint of implication that I am involved in whatever you are doing.

The infrastructure already exists, otherwise the Flathub community wouldn't able to maintain the 1000+ apps at the time of this writing. MultiMC isn't going to be much different.

**Comment by damianatorrpm** [comment]
Note: I am not a fltahub/flatpak decider, neither a Minecroft user.
@peterix I've read in to this discussion and I strongly disagree with your attitude.
The reason is that there are flatpak-only Linux distributions out and in the works.
You're actively saying as an open source developer: "I don't want my software to work on distribution XY".
You're offered to maintain the distribution of the flatpak yourself (and already do so for your preferred systems that you 
obviously try to push - like anbox does with snap and Nobuntu).
The edits of MultiMC in questions are configure path's basically, so your comment on changes can be only seen as a pretext.

So the following should resolve the question to determine you're true colors:
If you're really 'concerned' about the 5 lines of patch that edits MultiMc (the configure/java path)
than the following should have your blessing: A flatpak MultiMC not build from source but uses wine (https://winepak.org/)
to run that software unmodified.

**Comment by peterix** [comment]
I do not approve this.

I said NO many times.

And I see this continuing as hostile action. I will respond in kind if you force me.

**Comment by peterix** [comment]
I'd liken this to rape, actually.

**Comment by damianatorrpm** [comment]
@peterix You said no to a "modified" version, a version that is exactly re-packaged from a package from your website without modification is something completely different, so you must be okay with that, unless you're completely hostile towards any other distribution than arch and ubuntu.

**Comment by peterix** [comment]
No, I am against you repackaging or hosting MultiMC in any way. MultiMC comes from multimc.org - nowhere else.

**Comment by peterix** [comment]
And that is final.

**Comment by peterix** [comment]
If I have to make the repositories private and change the license, I will.

**Comment by peterix** [comment]
You people are entirely inconsiderate, relentless and can't be reasoned with in any way.

**Comment by damianatorrpm** [comment]
>You people are entirely inconsiderate, relentless and can't be reasoned with in any way.
So far nothing has been published. You have actually said moments ago that you do not wish this to be redistributed 
in any way (even without modification), yet you chose the Apache license. 
Maybe you should consider and appropriate license for your project if you don't wish that to happen and avoid all the drama.

**Comment by peterix** [comment]
Maybe you should consider 'no' as an answer and have some respect for other people.

**Comment by damianatorrpm** [comment]
I can imagine that this has been as bad of a discussion for you as for anyone involved, which you **could have avoided entirely**.
I've forked MultiMC on github, not that I am willing to put any effort into something that I have no use for.

You probably didn't know but on flatpak irc people are very compassionate to not include MultiMC based on your wishes.

If you wish to use the label open source, I would suggest that you also respect the open source initiative 
https://opensource.org/ and not label it as such and than disrespect people who than put work into distributing it according to that very license.

**Comment by peterix** [comment]
Great. That is the desired outcome. I do not want MultiMC to be in flatpak, appimage, snap or anything similar.
I do not want to deal with users of these things.

You could have avoided this by not being aggressive and trying to take over everything.

**Comment by damianatorrpm** [comment]
The desired outcome would definitely include also you not misrepresenting the open source brand.

**Comment by peterix** [comment]
Go read the license again. It gives you no rights to the name.

**Comment by damianatorrpm** [comment]
I suggest you do the same:
`2. Grant of Copyright License. Subject to the terms and conditions of this License, each Contributor hereby grants to You a perpetual, worldwide, non-exclusive, no-charge, royalty-free, irrevocable copyright license to reproduce, prepare Derivative Works of, publicly display, publicly perform, sublicense, and distribute the Work and such Derivative Works in Source or Object form.`

**Comment by peterix** [comment]
> 6. Trademarks. This License does not grant permission to use the trade names, trademarks, service marks, or product names of the Licensor, except as required for reasonable and customary use in describing the origin of the Work and reproducing the content of the NOTICE file.****

**Comment by damianatorrpm** [comment]
>except as required for reasonable and customary use

which is covered by section 2

**Comment by peterix** [comment]
No, it is not. It's a distinct section of the license.

**Comment by damianatorrpm** [comment]
@peterix Would you mind if we ask the OSI for clarification?

**Comment by peterix** [comment]
Look, I'm telling you 'no, I don't want this' and you are trying to find some technicality to get around it and ignore what I say.

**Comment by damianatorrpm** [comment]
I'm sorry but I don't want to get around anything, it's the other way around.

You are basically saying open source (apache 2) is not redistributable based on some technicality which is covered by section 2.
I'm merely asking you to chose the right license for your project to resolve current and future discussions.

It seems to me that you've chosen the incorrect license for the project (an open source one). 
You can't have it both ways. Either you are open source or you are not. (And both is fine!)
The open source brand is controlled by the OSI so I think we should ask them if you are allowed to use their brand (open source)
or not

**Comment by peterix** [comment]
The license does not give you the right to use the name or the infrastructure MultiMC requires to function.

There are two very distinct things:
MultiMC, the launcher as distributed on multimc.org - I am willing to support it and have for almost 9 years. It is packaged in a specific way and you don't get to say how.
MultiMC, the open source code on github.

It is clear that I need to strip all branding from the open source code and make it present only in binary packages built by the MultiMC CI.

**Comment by damianatorrpm** [comment]
>It is clear that I need to strip all branding from the open source code and make it present only in binary packages built by the MultiMC CI.

I think this is a solution and glad that you can be reasoned with. 
I hope that this will also prevent you from having to lead such discussion in the future, as they are not pleasant for anyone involved.

**Comment by barthalion** [comment]
People interested in having unbranded MultiMC on Flathub are welcome to submit it; at this point I don't want MultiMC on Flathub either.

**Pull request closed without merging**
